django/core/signing.py

```diff
@@ -166,7 +180,10 @@ def unsign(self, signed_value):
         if self.sep not in signed_value:
             raise BadSignature('No "%s" found in value' % self.sep)
         value, sig = signed_value.rsplit(self.sep, 1)
-        if constant_time_compare(sig, self.signature(value)):
+        attempts = [
+            constant_time_compare(sig, self.signature(value, key)) for key in self.keys
```

Comment:
We are doing a lot of work here when usually only the first key should be needed. I understand that usually the list will for short; but still…

what about:
```
for key in self.keys:
  valid = constant_time_compare(sig, self.signature(value, key))
  if valid:
    return value
```

Reply:
Is there any risk inherent in having multiple comparisons though?

We call `constant_time_compare()` intentionally to reduce the risk of timing attacks. If it were possible to trigger something to be unsigned against `SECRET_KEYS[0]` (one key used) and something against `SECRET_KEYS[1]` (two keys used), does that pose a problem? If anything we ought to *always* compare against all keys. Perhaps we should add a comment here to ensure that happens and a warning in the docs for `SECRET_KEYS` that all keys need to be checkd, so don't add too many?

Reply:
The "risk" is that it would add a constant overhead to every signing operation if there is more than one entry in the list. I currently cannot imagine an attack where knowing which of the secret keys (index number) was used -- but I'll happily be proven wrong.

Adding a doc comment would somewhat alleviate my concern since it would make it clear that `keys[1:]` is basically just transitional and should be removed asap. That said it simply feels wrong to perform unneeded work -- I know that this code is security sensitive and I will probably regret that I said that this work is unneeded when one indeed finds a security issue in the future… but still :)

Reply:
> The "risk" is that it would add a constant overhead to every signing operation if there is more than one entry in the list. I currently cannot imagine an attack where knowing which of the secret keys (index number) was used -- but I'll happily be proven wrong.

Me neither, but thought it worth raising for consideration.

> Adding a doc comment would somewhat alleviate my concern since it would make it clear that `keys[1:]` is basically just transitional and should be removed asap.

+1 to that.

> That said it simply feels wrong to perform unneeded work

Oh, I'd be the first to agree with that.

> I know that this code is security sensitive and I will probably regret that I said that this work is unneeded when one indeed finds a security issue in the future… but still :)

😅

Reply:
Fair warning, I don't know nearly enough about things of this nature to be actually valuable, except possibly to ask questions which might be dumb or might be previously unconsidered. So please take it with a grain of salt...

If an adversary is making use of the fact a secret key has been compromised, which scenario gives them less information when the key is subsequently discovered and rotated and they want to stop trying to make use of it (to fly under the radar as much as possible); one constant time compare, or multiple? Layman's gut feeling says you'd want the time comparison to be actually roughly constant, which would probably require always doing either 1 or 2 (I think, it's murky in my brain now) always, regardless of number of keys in the corpus? (ie: if there's only one key, always compare as if there are 2...)

And does that kind of scenario _actually_ matter? IDK.